src/audio/detect_test.c

```diff
@@ -0,0 +1,654 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+ */
+
+#include <stdint.h>
+#include <stddef.h>
+#include <sof/lock.h>
+#include <sof/list.h>
+#include <sof/stream.h>
+#include <sof/ipc.h>
+#include <sof/notifier.h>
+#include <sof/audio/component.h>
+#include <sof/audio/kpb.h>
+#include <uapi/user/detect_test.h>
+
+/* tracing */
+#define trace_keyword(__e, ...) \
+	trace_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+#define trace_keyword_error(__e, ...) \
+	trace_error(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+#define tracev_keyword(__e, ...) \
+	tracev_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+
+#define ACTIVATION_DEFAULT_SHIFT 3
+#define ACTIVATION_DEFAULT_THRESHOLD 0.5
+
+#define ACTIVATION_DEFAULT_THRESHOLD_S16 \
+	((int16_t)((INT16_MAX) * (ACTIVATION_DEFAULT_THRESHOLD)))
+
+/* number of samples to be treated as a full keyphrase */
+#define KEYPHRASE_DEFAULT_PREAMBLE_LENGTH (30 * 1024)
+
+struct comp_data {
+	uint32_t period_bytes;		/**< source number of period bytes */
+
+	struct sof_detect_test_config config;
+	void *load_memory;	/**< synthetic memory load */
+	int16_t activation;
+	uint32_t detected;
+	uint32_t detect_preamble; /**< current keyphrase preamble length */
+	uint32_t keyphrase_samples; /**< keyphrase length in samples */
+	uint32_t buf_copy_pos; /**< current copy position for incoming data */
+
+	struct notify_data event;
+	struct kpb_event_data event_data;
+	struct kpb_client client_data;
+
+	void (*detect_func)(struct comp_dev *dev,
+			    struct comp_buffer *source, uint32_t frames);
+};
+
+static void notify_host(struct comp_dev *dev)
+{
+	struct sof_ipc_comp_event event;
+
+	trace_keyword("notify_host()");
+
+	event.event_type = SOF_CTRL_EVENT_KD;
+	event.num_elems = 0;
+
+	ipc_send_comp_notification(dev, &event);
+}
+
+static void notify_kbd(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("notify_kbd(), preamble: %u", cd->detect_preamble);
+
+	cd->client_data.r_ptr = NULL;
+	cd->client_data.sink = NULL;
+	cd->client_data.id = 0; /**< TODO: acquire proper id from kpb */
+	cd->client_data.history_end = 0; /**< keyphrase end, 0 is now */
+	cd->client_data.history_begin = cd->detect_preamble;
+	/* two channels */
+	cd->client_data.history_depth = cd->detect_preamble *
+		sizeof(int16_t) * 2;
+
+	cd->event_data.event_id = KPB_EVENT_BEGIN_DRAINING;
+	cd->event_data.client_data = &cd->client_data;
+
+	cd->event.id = NOTIFIER_ID_KPB_CLIENT_EVT;
+	cd->event.target_core_mask = NOTIFIER_TARGET_CORE_ALL_MASK;
+	cd->event.data = &cd->event_data;
+
+	notifier_event(&cd->event);
+}
+
+static void detect_test_notify(struct comp_dev *dev)
+{
+	notify_host(dev);
+	notify_kbd(dev);
+}
+
+static void default_detect_test(struct comp_dev *dev,
+				struct comp_buffer *source, uint32_t frames)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	int16_t *src = source->r_ptr;
```

Comment:
Pointers and frames may not be aligned to buffer size, so use buffer_read_frag_s16.